Added GBA/DS reverse engineering tutorials (#20)
StarcubeLabs/StarcubeLabs.github.io: _posts/2023-05-06-reverse-engineering-gba.markdown

---
layout: post
title: Reverse Engineering a GBA Game
date: 2023-05-06 05:00:00
img: reverse-engineering/debug-env.png

> description: "How to reverse engineer a GBA game, using Pokémon Mystery Dungeon: Red Rescue Team as an example."

Reviewer: ```suggestion
description: "How to reverse engineer a GBA game using \"Pokémon Mystery Dungeon: Red Rescue Team\" as an example."
```

Author: Good call, though I've seen italics more than quotes for game names so I'll do that.

author: Some Body
---

## Intro

> This tutorial will introduce the process of reverse engineering a game. We'll use the Game Boy Advance game _Pokémon Mystery Dungeon: Red Rescue Team_ for demonstration purposes, although most concepts discussed here can be applied to other games.

Reviewer: ```suggestion
This tutorial introduces the process of reverse engineering a game. We'll use the Game Boy Advance game _Pokémon Mystery Dungeon: Red Rescue Team_ for demonstration purposes, although most concepts discussed here can be applied to other games.
```

### What is this tutorial about?
This tutorial will walk you through the following:
* Setting up a reverse engineering environment.
* A basic introduction to assembly code.

> * Using the reverse engineering tool Ghidra to analyze assembly code.

Reviewer: ```suggestion
* Using the reverse engineering tool \"Ghidra\" to analyze assembly code.
```

Author: I feel quotes aren't needed here. It's common to use a description followed by a name.

> * Exploring emulator debugging features to inspect a game's memory and debug the game while it is running.

Reviewer: ```suggestion
* Exploring debugging features of the gba emulator to inspect a game's memory and debug the game while it is running.
```

Author: I'll name-drop mGBA as the specific GBA emulator since I did the same for Ghidra.

4. _File > Import File..._
5. Select the Red Rescue Team ROM on your file system.
6. Select an instruction set architecture for Ghidra to use for analyzing the binary. The GBA uses the ARMv4T instruction set in little-endian format (__ARM:LE:32:v4t__), which you can find by searching for "v4t" in the language select, then selecting the option with "little" under the _Endian_ column.
    >![](/assets/img/reverse-engineering/armv4t.png)<br>
    >Selecting the ARMv4T language
7. Click _Options..._ and set the Base Address to 0x8000000.
7. Click _OK_ to make Ghidra inspect the binary.
8. Once done, Ghidra will display a report with some details about the inspection. Click _OK_ to continue.
9. Double-click the ROM in Ghidra to open Ghidra's code viewer.
10. Ghidra will offer to analyze the binary. Click _Yes_.
11. Use the default analysis settings and click _Analyze_.
12. Wait for Ghidra to analyze the binary. This may take a couple minutes.
13. When Ghidra's analysis finishes, it will present a screen like the image below. At this point, Ghidra is set up to start reverse engineering the game.
>![](/assets/img/reverse-engineering/ghidra-setup.png)<br>
>Ghidra after setting it up

> How would you find that you should use the ARM:LE:32:v4t language if I didn't tell you? First search for which CPU the GBA uses; you'll find that it uses the ARM7TDMI processor (e.g., from [Wikipedia](https://en.wikipedia.org/wiki/Game_Boy_Advance)). Knowing the CPU, you can look for which language/architecture the CPU uses; [ARM's documentation](https://developer.arm.com/documentation/ddi0210/c/Introduction/Architecture) states that the ARM7TDMI uses the ARMv4T architecture.

Reviewer: ```suggestion
How would you know to use the ARM:LE:32:v4t language if I didn't tell you? First search for which CPU the GBA uses; you'll find that it uses the ARM7TDMI processor (e.g., from [Wikipedia](https://en.wikipedia.org/wiki/Game_Boy_Advance)). Knowing the CPU, you can look for which language/architecture the CPU uses; [ARM's documentation](https://developer.arm.com/documentation/ddi0210/c/Introduction/Architecture) states that the ARM7TDMI uses the ARMv4T architecture.
```

> As for [endianness](https://en.wikipedia.org/wiki/Endianness) (the order that bytes are stored for each [word](https://en.wikipedia.org/wiki/Word_(computer_architecture)) of data), most ARM CPUs are "bi-endian" and support both little and big endian, but in practice most ARM programs use little endian. If Ghidra spits out incomprehensible disassembled code when using a little endian language, try big endian and see if the output is any better. Note that some CPUs, such as those using the [PowerPC](https://en.wikipedia.org/wiki/PowerPC) instruction set (e.g., the Wii), are predominantly big endian instead.

Reviewer: ```suggestion
As for [endianness](https://en.wikipedia.org/wiki/Endianness) (the order that bytes are stored for each [word](https://en.wikipedia.org/wiki/Word_(computer_architecture)) of data), most ARM CPUs are "bi-endian" and support both little and big endian. In practice most ARM programs use little endian. If Ghidra spits out incomprehensible disassembled code while using a little endian language, try big endian and see if the output is any better. Note that some CPUs, such as those using the [PowerPC](https://en.wikipedia.org/wiki/PowerPC) instruction set (e.g., the Wii), are predominantly big endian instead.
```

```
adds r0,#0x1
```
This will add 1 to the value of `r0` and store the sum in `r0`. Alternatively, you can say that this increments the value in `r0`.

Other available math operators include subtraction, multiplication, negation, bitwise AND/OR/XOR/NOT, and logical/arithmetic (unsigned/signed) left/right bit shifts. Some of these operations are more rigid than addition, such as not supporting immediate values, though they all follow a similar structure. For a full list of supported instructions, refer to [ARM's developer documentation](https://developer.arm.com/documentation/ddi0210/c/Introduction/Instruction-set-summary/Thumb-instruction-summary?lang=en).

>You may notice the lack of a division instruction. Division with arbitrary numbers is more complicated than the operations listed above, so it is implemented as a function rather than a single instruction. This means division is much slower than other math operations. Note that the right shift operator can divide numbers by powers of 2 to achieve certain divisions in a single instruction.

#### Load/Store memory

> Registers can only store a handful of values, so memory is used to store the majority of a program's data.

Reviewer: I would add a blurb here that memory here is RAM and it is where the remainder of the program values are stored.

Registers `r0`-`r3` are designated as __scratch registers__, which are not saved by a function that uses them. Register `r4`-`r11` are __preserved registers__ whose values are saved and restored by the function. `lr` is also saved if the function calls other functions.

Since a function can call another function, which can itself call another function and so on, every function must store and restore the register values at the proper times. This is accomplished using a place in memory called the call stack.

The __call stack__, often shortened to __stack__, is a special place in memory used to save register values when functions are called. It is also used to store local variables if there are too many for the registers to hold. As the name implies, it is a last-in first-out (LIFO) data structure.

In most games, the stack is located in memory at addresses 0x3000000-0x3007FFF. The top of the stack starts at address 0x3007FFF at the beginning of program execution. As values are pushed onto the stack, the top of the stack moves down. The address at the top of the stack is tracked using `r13`, often called the __stack pointer__ (`sp`).

> One of the main purposes of the function prologue is to save register values to the stack. Register values are pushed onto the top of the stack (i.e., the address of `sp`) using the `push` instruction. The instruction also decrements `sp` to reflect the new top of the stack.

Reviewer: This is the first time a function prologue is mentioned and I'm not sure what it is.

Author: The parts of the function, including the prologue, are introduced on line 373.

>![](/assets/img/reverse-engineering/debugger-read.png)<br>
>Reading a memory value

Note that most debugger console commands will pause the game if it is not already paused.

Writing values to memory uses the format `w/<size> <address> <write value>`. For instance, the command `w/1 0x200419E 20` will set the player's HP to 20. Like the address, the write value can be either decimal or hex. There is also a `w/r` command to write a value to a register instead of memory.

#### Breakpoints
Similar to debuggers in higher-level languages, a breakpoint pauses program execution if a specific line of code is reached. This has several uses, like figuring out whether a line of code is reached when performing an action in-game, or stopping program execution at a specific function to debug it.

A breakpoint can be set using the `b` (`break`) command, with the format `b <address>`. As a demonstration, set a breakpoint at the beginning of the function `FUN_08044b28` using `b 0x08044b28`. This will pause game execution, so unpause the game with the `c` command.
>![](/assets/img/reverse-engineering/debugger-add-breakpoint.png)<br>
>Adding a breakpoint

> With the breakpoint added, try to move around in-game. This will execute `FUN_08044b28` and hit the breakpoint, pausing the game's execution.

Reviewer: ```suggestion
With the breakpoint added, try to moving your character around in the gba emulator. This will execute `FUN_08044b28` and hit the breakpoint, pausing the game's execution.
```

Author: The term "emulator" includes both the game window and other supporting windows like the debugging console, so "in-game" is a more specific description. I'll mention the character though.

### mGBA wrap-up
We've seen some of the general tools in mGBA for inspecting and debugging the live game. The emulator has a number of more specialized tools that this tutorial won't go into, such as viewing the currently loaded palettes, sprites, and backgrounds. You may wish to check them out on your own if those strike your fancy.

## Reverse engineering strategies
With assembly knowledge, Ghidra, and mGBA, we have all the tools to start reverse engineering a game. The last section of this tutorial will discuss reverse engineering strategies to find where functionality lives in the ROM and in memory.

The mGBA section of this tutorial discusses a couple of strategies using [memory search](#memory-search) and [watchpoints](#watchpoints), so check those out if you haven't already.

### Following a value backwards through assembly
One way to find the location of a specific value is to trace related values backwards through the assembly to find where they came from. This might lead to the value you are looking for.

> To demonstrate, let's find the place in code that determines how much a certain healing item will heal. In-game, explore the dungeon floor until you find a blue berry, called an Oran Berry. You may not necessarily find one on the floor you're on; if you've explored the floor and found nothing, look for a staircase to proceed to the next floor and continue the search.

Reviewer: ```suggestion
To demonstrate, let's find the place in code that determines how many hit points a certain healing item will heal your character. In-game, explore the dungeon floor until you find a blue berry, called an Oran Berry. You may not necessarily find one on the floor you're on; if you've explored the floor and found nothing, look for a staircase to proceed to the next floor and continue the search.
```

The two previous lines of assembly are also of note.
```
mov r3,r10
add r0,r2,r3
```
The 0x65 in `r0` was the sum of `r2` and `r3`, and looking at the current game state from the watchpoint, `r2` matches the player's HP of 1 while `r3` matches the Oran Berry's healing amount of 0x64 (100). Prior to the addition, we can see that the 0x64 in `r3` was copied from `r10`.

Now let's look for where the value 0x64 was assigned to `r10`. Middle-click `r10` to highlight its usages, then scroll up in the function. There is another reference to `r10` at 0x77c8c with `mov r3,r10`, but this doesn't assign to `r10`, so we can ignore it. Near the top of the function is what we are looking for, address 0x77c54 with `mov r10,r2`.
>![](/assets/img/reverse-engineering/trace-oran-berry.png)<br>
>Tracing the healing amount of the Oran Berry to the top of the function

> The 0x64 in `r10` came from `r2`, but there are no more references to `r2` between here and the start of the function. This means that `r2` is the third parameter of `FUN_08077c44`, and the 0x64 came from the code that called `FUN_08077c44`. The next step is to find the caller of this function.

Reviewer: ```suggestion
The 0x64 in `r10` came from `r2`, but there are no more references to `r2` between here and the start of the function. This means that `r2` is the third parameter of `FUN_08077c44`, and the 0x64 came from the code that called `FUN_08077c44` (remember registers r0 through r3 are "scratch registers" used for function arguments by convention). The next step is to find the caller of this function.
```

>![](/assets/img/reverse-engineering/oran-berry-fixed-assembly.png)<br>
>0x804838D and surrounding addresses after manually disassembling it

Now the call to `FUN_08077cc44` is visible at address 0x8048388. We were looking for where `r2` was assigned to 0x64, so look at the lines prior to the function call. We can see this set of instructions starting at 0x8048378.
```
ldr r2,[DAT_08048394]
mov r3,#0x0
ldrsh r2,[r2,r3]
```
This loads an address into `r2` from data value `DAT_08048394`, then loads the value at the address into `r2`. Look at address 0x48394 to find that the value of `DAT_08048394` is 0x080F4FB6. This is within ROM, which is promising; ROM cannot normally be written to, so chances are that this is the source of the value 0x64.

With this address in mind, go back to mGBA and go to address 0x080F4FB6 in the memory viewer. You will find the value 0x64 at the address, confirming that this is the location of the Oran Berry's HP heal amount.
>![](/assets/img/reverse-engineering/oran-berry-found.png)<br>
>The location of the Oran Berry's heal amount

> To test the address we found for the heal amount, reload your save state, change the 0x64 at address 0x080F4FB6 to 0x0A (10 in decimal) in the memory viewer, and eat the Oran Berry again to see how much it heals. You may wish to delete the watchpoint and breakpoint from earlier to avoid hitting them again now that we are finished with them.

Reviewer: ```suggestion
To test the address we found for the heal amount, reload your save state, change the 0x64 at address 0x080F4FB6 to 0x0A ("0x0A" is 10 in decimal) in the memory viewer, and eat the Oran Berry again to see how much it heals. You may wish to delete the watchpoint and breakpoint from earlier to avoid hitting them again now that we are finished with them.
```

>![](/assets/img/reverse-engineering/oran-berry-changed.png)<br>
>The changed Oran Berry heal amount

> Success! We've found that the Oran Berry's heal amount is stored at 0x080F4FB6. Technically, if you edited the game binary and changed the value at byte 0xF4FB6, you'd create a minimal ROM hack that nerfs the Oran Berry's healing. Creating ROM hacks is out of the scope of this tutorial, but this demonstrates a piece of the process for creating these hacks.

Reviewer: ```suggestion
Success! We've found that the Oran Berry's heal amount is stored at 0x080F4FB6. Technically, if you edited the game binary and changed the value at byte 0xF4FB6, you'd create a minimal ROM hack that nerfs the Oran Berry's healing. Creating ROM hacks is out of the scope of this tutorial, but this demonstrates a piece of the process for creating these types of hacks.
```

>![](/assets/img/reverse-engineering/memory-viewer-nearby.png)<br>
>Looking at values near the player's HP in memory

A value directly related to your HP is your maximum HP. Look in the in-game HUD for your max HP (directly to the right of your current HP), then try to find it in the memory viewer.

You won't have to look very far. Two bytes ahead of your current HP is a number that matches your max HP, address 0x020041A0. Try to change this value, and you will see that your max HP changes on the HUD, confirming that 0x020041A0 is the player's max HP.

Another number that is close by is the player's level (labeled "Lv" on the HUD), which is 5 when you've just started the game. Look around the memory viewer for the value 5, and once you've found a possible value, try changing it to see if the player's level changes in the HUD. The correct address is in the spoiler below.
<details>
  <summary>Address of the player's level</summary>
  0x02004199
</details>

> A related approach is to make actions in-game and watch as values in memory change. This includes moving around, attacking, etc. To demonstrate this approach, press the Start button to enter a mode where you can face a direction without moving. Turn around a couple of times while watching the memory viewer, and you'll see the value at address 0x020041D6 changing. No other visible value changes, which is a good indicator that 0x020041D6 contains the direction the player was facing (encoded as an enum).

Reviewer: ```suggestion
A related approach is to perform an action in-game and watch as the values in memory change. This includes moving around, attacking, etc. To demonstrate this approach, press the Start button to enter a mode where you can face a direction without moving. Turn around a couple of times while watching the memory viewer, and you'll see the value at address 0x020041D6 changing. No other visible value changes, which is a good indicator that 0x020041D6 contains the direction the player was facing (encoded as an enum).
```
